e2e: ginkgo timeouts: use context provided by ginkgo

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Ginkgo can tell code to stop running by canceling the context that Ginkgo provides to callbacks when those accept one. This can be used to stop immediately when aborting manually via CTRL-C and to clean up properly in case of a timeout, because cleanup code then runs after the main test has stopped with a new context.

#### Special notes for your reviewer:

I started eliminating context.TODO in provisioning.go and then branched out from there: any function which had context.TODO needed an explicit context. The minimal goal for this PR is to clean up all code in test/e2e/framework because then further PRs probably can be more localized.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

aojea commented:
is this the one you are testing the new contexts?

pohly commented:
Yes.

pohly commented:
Test failures are intentional - I broke one test to see how timeouts get reported.

fedebongio commented:
/triage accepted

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-node-e2e-containerd

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce-storage-slow
/test pull-kubernetes-e2e-gci-gce-ingress

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce-storage-slow
/test pull-kubernetes-e2e-gci-gce-ingress

pohly commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-conformance-kind-ga-only-parallel

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-conformance-kind-ga-only-parallel

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-conformance-kind-ga-only-parallel

pohly commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-conformance-kind-ga-only-parallel
/test pull-kubernetes-node-e2e-containerd 
/test pull-kubernetes-e2e-gce-100-performance

pohly commented:
I am now at a point where `context.TODO` has been removed from all files that I touch in this PR, i.e. I've reached some kind of closure.

More work can be found with the grep commands in the last commit message, though, so further PRs would be needed if we merge this PR as-is.

pohly commented:
/retest pull-kubernetes-e2e-gce-ubuntu-containerd

pohly commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

pohly commented:
/test pull-kubernetes-conformance-kind-ga-only
/test pull-kubernetes-conformance-kind-ipv6-parallel
/test pull-kubernetes-e2e-capz-azure-disk
/test pull-kubernetes-e2e-capz-azure-disk-vmss
/test pull-kubernetes-e2e-capz-azure-file
/test pull-kubernetes-e2e-capz-azure-file-vmss
/test pull-kubernetes-e2e-capz-conformance
/test pull-kubernetes-e2e-capz-ha-control-plane
/test pull-kubernetes-e2e-capz-windows-containerd
/test pull-kubernetes-e2e-containerd-gce
/test pull-kubernetes-e2e-gce-alpha-features
/test pull-kubernetes-e2e-gce-correctness
/test pull-kubernetes-e2e-gce-csi-serial
/test pull-kubernetes-e2e-gce-device-plugin-gpu
/test pull-kubernetes-node-e2e-containerd-alpha-features
/test pull-kubernetes-node-e2e-containerd-features
/test pull-kubernetes-node-e2e-containerd-features-kubetest2
/test pull-kubernetes-node-e2e-containerd-kubetest2
/test pull-kubernetes-node-kubelet-credential-provider
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-kubetest2
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-cpu-manager-kubetest2
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv1
/test pull-kubernetes-node-kubelet-serial-crio-cgroupv2
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager-kubetest2

spiffxp commented:
/cc
Looks this builds on top of https://github.com/kubernetes/kubernetes/pull/113367

spiffxp commented:
/test pull-kubernetes-node-e2e-containerd

spiffxp commented:
This is gigantic.  I'm making my through it assuming it is nearly ready to go, but the number of job failures is a little concerning.  At a minimum pull-kubernetes-node-e2e-containerd failing needs to be addressed (haven't had time to look into why it's failing but it failed a retest)

Is there any part of this you could break down so it doesn't get caught in rebases?

spiffxp reviewed (commented):
I'm not going to be around to re-review this tomorrow. In principal I'm OK with these changes, I just think it's a lot to do all at once. Given that code freeze / test freeze are coming up I can't help but wonder if there's a way to shard these up or break into more commits for easier review.

pohly commented:
> This is gigantic. I'm making my through it assuming it is nearly ready to go, but the number of job failures is a little concerning.

That's partly because I picked several pre-merge jobs and manually ran them although they are normally not started. Several of those jobs seem to be just plain broken, independently of this PR.

> At a minimum pull-kubernetes-node-e2e-containerd failing needs to be addressed (haven't had time to look into why it's failing but it failed a retest)

I'll check.

> Is there any part of this you could break down so it doesn't get caught in rebases?

Not really. Once I started requiring a context parameter in e2e/framework, it bubbled up into pretty much everything. I might have been able to stop at some point half the way, but I can't get back to that state anymore.

pohly commented:
One possibility would be to wait with merging this until after code freeze and before test freeze. Then we don't make everyone else's life miserable because their new tests no longer work while they try to hit code freeze. We may still have conflicts when new tests get added after code freeze, but probably less than now.

Rebasing isn't too bad - I just take whatever chunk comes from master and fix the compile errors.

pohly commented on test/e2e/framework/test_context.go:
```diff
@@ -539,7 +539,9 @@ func writeJUnitReport(report ginkgo.Report) {
 		// information, so we don't need it in the XML.
-		if specReport.State != types.SpecStateFailed {
+		switch specReport.State {
+		case types.SpecStateSkipped, types.SpecStatePassed:
```
Sorry, old habits... fixed.

pohly commented on test/e2e/framework/ginkgowrapper.go:
```diff
@@ -18,15 +18,47 @@ package framework
+// `apierrors.IsNotFound` considers as "not found", the error is ignored
+// instead of failing the test during cleanup. This is useful for cleanup code
+// that just needs to ensure that some object does not exist anymore.
+func IgnoreNotFound(in any) any {
```
I can add it in a separate commit with a better explanation why it is needed: several `ginkgo.DeferCleanup` calls invoke a delete operation and are not supposed to fail when the object that needs to be deleted is already gone.

pohly replied:
Actually, I already added it in `e2e framework: support ignoring "not found" errors during DeferCleanup`, but that commit is not as clean as it should be. I'll fix that.

aojea replied:
this is super complex :/ 
do we have many of these failures?

pohly replied:
There were a few, usually coming from tests that had a `defer func() { ... Delete ... }` without return code checking when those tests also deleted the object as part of normal testing.

pohly commented on test/e2e/framework/framework.go:
```diff
@@ -268,6 +268,9 @@ func (f *Framework) dumpNamespaceInfo() {
 	if !TestContext.DumpLogsOnFailure {
 		return
 	}
+	if f.DumpAllNamespaceInfo == nil {
+		return
+	}
```
This needs to be in its own commit. It fixes a nil pointer access in the output unit test where DumpAllNamespaceInfo isn't set.

aojea replied:
yeah, I hit this the other day here https://github.com/kubernetes/kubernetes/pull/113850

pohly replied:
There are several AfterEach callbacks that use variables set in BeforeEach or during tests, without checking that the initialization actually completed. Using DeferCleanup right after initialization avoids this pitfall, we just can't use it here and thus have to check.

pohly commented:
The  pull-kubernetes-node-e2e-containerd was caused by a misplaced `ctx` parameter in a `ginkgo.Context` callback. This is not valid, which caused the suite to fail during startup. All of those container node calls get executed, so it's not possible to miss such a mistake.

logicalhan commented:
/assign @dashpole 
/triage accepted

spiffxp commented:
>One possibility would be to wait with merging this until after code freeze and before test freeze.

That sounds reasonable, ping once you've rebased after code freeze

pohly commented:
/retest

aojea commented on test/e2e/storage/drivers/csi-test/driver/driver.go:
```diff
@@ -157,7 +157,7 @@ func goServe(server *grpc.Server, wg *sync.WaitGroup, listener net.Listener, sta
 		started <- true
 		err := server.Serve(listener)
 		if err != nil {
-			panic(err.Error())
```
it seems that this should be using the `defer ginkgo.GinkgoRecover()`

pohly replied:
That would have made the panic show up as a Ginkgo failure instead of just being recorded in the log output. Panics shouldn't happen in this goroutine, therefore I am not sure whether it's worth it.

aojea commented on hack/conformance/check_conformance_test_requirements.go:
```diff
@@ -30,7 +30,7 @@ import (
 
 const (
-	//e.g. framework.ConformanceIt("should provide secure master service ", func() {
+	//e.g. framework.ConformanceIt("should provide secure master service ", func(ctx context.Context) {
```
why don't you add a `hack/verify-ginkgo-context.sh`and `hack-update-ginkgo-context.sh` so we are consistent with this ?
It also will give us more confident in this change having the scripts in repo

pohly replied:
Not all callbacks need a context, only those that call functions which need a context. In practice, almost all tests will need it because they communicate with a cluster, which implies that they could block and thus should use a context. But should we make it a requirement for all tests?

There may be some value in it. Developers might be tempted to write a test like this because they don't know better:
```
ginkgo.It("...", func() {
    ctx := context.TODO() // For framework helper functions below.
   ...
    e2epod.WaitForPod(ctx, ...)
})
```

The sed invocation isn't perfect, though. A few tests were defined with
```
ginkgo.It("...",
   func() {
   }
)
```

Because sed expected the `func` on the same line as `ginkgo.It`, they weren't updated. Should the scripts perhaps enforce that they are on the same line? Tests that use two lines need to indent the function body one level deeper than tests that use one line, and it is inconsistent.

pohly replied:
I just checked, there are some places where sed is adding context parameters unnecessarily:
```
diff --git a/test/e2e/dra/dra.go b/test/e2e/dra/dra.go
index fc7529e682c..24aa9301249 100644
--- a/test/e2e/dra/dra.go
+++ b/test/e2e/dra/dra.go
@@ -61,7 +61,7 @@ var _ = ginkgo.Describe("[sig-node] DRA [Feature:DynamicResourceAllocation]", fu
                nodes := NewNodes(f, 1, 1)
                driver := NewDriver(f, nodes, networkResources) // All tests get their own driver instance.
                b := newBuilder(f, driver)
-               ginkgo.It("registers plugin", func() {
+               ginkgo.It("registers plugin", func(ctx context.Context) {
                        ginkgo.By("the driver is running")
                })
```

Here all the relevant work was done in `BeforeEach`, the test just exists to have something simple which exercises that and nothing else.

The other cases are `test/e2e/framework/log_test.go` and `test/e2e/framework/skipper/skipper_test.go` which are simpler when the callbacks don't use a context.

All of these could be converted, I just wanted to share my observation.

aojea replied:
Tim wanted to have this https://github.com/quasilyte/go-ruleguard that will allow us to write rules IIUIC :( ... but yeah, it has to be something more fancy that sed, it should understand the function signature

pohly replied:
I remember that one. I think we may be able to use it through golangci-lint because there is one plugin which accepts ruleguard configuration, but I haven't checked.

I would prefer to postpone enforcement of new rules until the dust has settled. Let's do that for 1.27.

aojea replied:
sure, it was a side comment, didn't mean to start fiddling with that now

aojea commented on test/e2e/apimachinery/apiserver_identity.go:
```diff
@@ -37,7 +37,7 @@ import (
 )
 
-func getControlPlaneHostname(node *v1.Node) (string, error) {
+func getControlPlaneHostname(ctx context.Context, node *v1.Node) (string, error) {
```
I suggest to merge this commit in a follow up PR, so we have some time to observe the behavior of the CI.

pohly replied:
"This commit" = "[e2e: use context from Ginkgo](https://github.com/kubernetes/kubernetes/pull/112923/commits/8838fc4314822ae079132486bb0c056c8ff59e3a)"?

But then the rest of the changes are pretty harmless (I know, famous last words...). How long would you want to test those before doing the real change?

aojea replied:
some soak time, that is better than doing a full revert later or start bisecting ... it seems that if we do this today we have to merge the other tomorrow https://www.kubernetes.dev/resources/release/

pohly replied:
I've split out the "easy" parts into [a separate PR](https://github.com/kubernetes/kubernetes/pull/113884) and rebased this one here on top of it so that the complex commit is at the end.

We can first merge that and then this one (they won't conflict), or go directly for this one.

aojea replied:
defer to @spiffxp

pohly commented:
/retest

aojea commented on test/e2e/apimachinery/aggregator.go:
```diff
@@ -57,38 +57,24 @@ const (
 	// after NewDefaultFramework.
 	ginkgo.BeforeEach(func() {
-		c = f.ClientSet
-		ns = f.Namespace.Name
-
-		if aggrclient == nil {
-			config, err := framework.LoadConfig()
-			if err != nil {
-				framework.Failf("could not load config: %v", err)
-			}
-			aggrclient, err = aggregatorclient.NewForConfig(config)
-			if err != nil {
-				framework.Failf("could not create aggregator client: %v", err)
-			}
```
I thought already commented, this change is unrelated to the contexts, right? can we split it out?

pohly replied:
It's related because `ginkgo.DeferCleanup` can handle functions which take a context, i.e. this change became necessary when cleanTest was changed to require a contect.

It can be split out, but then it has to get merged before this PR.

pohly replied:
=> https://github.com/kubernetes/kubernetes/pull/114392

aojea commented on test/e2e/apimachinery/flowcontrol.go:
```diff
@@ -313,7 +310,7 @@ func getPriorityLevelNominalConcurrency(c clientset.Interface, priorityLevelName
 // createFlowSchema creates a flow schema referring to a particular priority
 // level and matching the username provided.
-func createFlowSchema(f *framework.Framework, flowSchemaName string, matchingPrecedence int32, priorityLevelName string, matchingUsernames []string) (*flowcontrol.FlowSchema, func()) {
+func createFlowSchema(ctx context.Context, f *framework.Framework, flowSchemaName string, matchingPrecedence int32, priorityLevelName string, matchingUsernames []string) *flowcontrol.FlowSchema {
```
`createPriorityLevel` has the same signature, however, you choose to use `ginkgo.DeferCleanup(` here instead of returning a `cleanup` function , I think that is better to be consistent, maybe avoid the ginkgo primitives by now?

pohly replied:
Using `ginkgo.DeferCleanup` was often the right solution for dealing with contexts during test cleanups. Not using it would have been more work and the outcome would have been more complicated. I had two choices:
- change the cleanup function to accept a context, which also means changing the caller
- eliminate the cleanup function

aojea replied:
but in some places the context is provided in the constructor, so the closure should inherit it , no?

pohly replied:
Each context from Ginko is only valid in the function that it is passed into. Using the context of the `BeforeEach` or `It` also for the cleanup operation is a bug. That should never be done because the API calls in the cleanup operation then abort immediately with "context cancelled" - one reason why it is important to check the error of those calls.

aojea commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -1727,21 +1699,19 @@ func registerWebhookForCustomResource(f *framework.Framework, configName string,
-	return func() {
-		client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.TODO(), configName, metav1.DeleteOptions{})
-	}
+	ginkgo.DeferCleanup(framework.IgnoreNotFound(client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete), configName, metav1.DeleteOptions{})
```
these are the behavioural changes I'd prefer to not do at the same time, this wasn't asserting any error before ... maybe we can avoid all these `ginkgo.DeferCleanup` changes and do only the contexts in this commit?

pohly replied:
Deferring the usage of `ginkgo.DeferCleanup` means that all of these lines must be touched twice.

pohly replied:
Let me try one thing: I'll do a pass through the PR and extract those parts that introduce `ginkgo.DeferCleanup`. It should be possible to merge those parts first.

It will take time, though, because in several places I'll then conflict with my own follow-up changes.

aojea replied:
that will be great, thanks, just the context only are straightforward to review and approve ...

aojea commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -2166,19 +2132,25 @@ func registerSlowWebhook(f *framework.Framework, configName string, certCtx *cer
 
-	return func() {
-		client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(context.TODO(), configName, metav1.DeleteOptions{})
+	cleanup := func(ctx context.Context) {
+		err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Delete(ctx, configName, metav1.DeleteOptions{})
+		if !apierrors.IsNotFound(err) {
+			framework.ExpectNoError(err)
+		}
 	}
```
it is more verbose but it is easier to understand that the internals of `framework.IgnoreNotFound(` using reflection and type assertion

pohly committed:
e2e: use Ginkgo context

All code must use the context from Ginkgo when doing API calls or polling for a
change, otherwise the code would not return immediately when the test gets
aborted.

dims commented:
If you still need this PR then please rebase, if not, please close the PR

pohly committed:
e2e storage: pass around topologyTest pointer

Copying the instance makes no sense because the fields are meant to be shared
between test and cleanup code.

pohly commented:
/retest

Unrelated flake: https://storage.googleapis.com/k8s-triage/index.html?test=Pods%20Extended%20Pod%20Container%20lifecycle%20should%20not%20create%20extra%20sandbox%20if%20all%20containers%20are%20done

pohly commented:
pull-kubernetes-e2e-gci-gce-ingress, "LoadBalancers should be able to preserve UDP traffic when server pod cycles for a LoadBalancer service on the same nodes" is an unrelated test flake:

https://storage.googleapis.com/k8s-triage/index.html?text=Failed%20to%20hit%20the%202%20UDP&test=LoadBalancers%20should%20be%20able%20to%20preserve%20UDP%20traffic%20when%20server%20pod%20cycles%20for%20a%20LoadBalancer%20service%20on%20the%20same%20nodes

pohly commented:
/retest

pohly commented:
/retest

"pull-kubernetes-e2e-capz-windows-containerd — Pod pending timeout"

aojea commented:
I tried my best to get to all files but there are 418 files changes, that is impossible, most of them are mechanical changes s/context.TODO/context gingo/ and we can always revert.


I think that merging on weekend will give us some soak time to see if something breaks, so we can revert in that case

/lgtm
/approve

e2e framework should be able to stop cleanly

onsi commented:
🙌
